File: addon/src/nodes/color_tint_decal.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright © 2026 The Halo Archive
from typing import cast
import bpy
from bpy.types import (
    NodeGroupInput,
    NodeGroupOutput,
    NodeSocketColor,
    NodeSocketFloat,
    NodeSocketShader,
    NodeTree,
    ShaderNodeBsdfPrincipled,
    ShaderNodeBsdfTransparent,
    ShaderNodeMath,
    ShaderNodeMixShader,
    ShaderNodeNewGeometry, ShaderNodeNormalMap, ShaderNodeGroup, ShaderNodeTree, ShaderNodeMix,
)

from .norm_normalize import NormNormalize
from ..utils import create_node, create_socket, create_link, assign_value

__all__ = ["ColorTintDecal"]


class ColorTintDecal:
    def __init__(self) -> None:
        self.node_tree: NodeTree | None = bpy.data.node_groups.get("Color Tint Decal Shader")
        if self.node_tree:
            return
        else:
            self.node_tree = bpy.data.node_groups.new(
                type="ShaderNodeTree", name="Color Tint Decal Shader"
            )
        self.create_sockets()
        self.create_nodes()

    def create_sockets(self) -> None:
        if self.node_tree is None:
            return
        interface = self.node_tree.interface
        _ = create_socket(interface, "BSDF", NodeSocketShader, False)
        _ = create_socket(interface, "Color Texture", NodeSocketColor)
        _ = create_socket(interface, "Color Alpha", NodeSocketFloat)
        _ = create_socket(interface, "Normal Map", NodeSocketColor)
        _ = create_socket(interface, "Roughness", NodeSocketFloat)
        _ = create_socket(interface, "Opacity", NodeSocketFloat)
        _ = create_socket(interface, "Metallic", NodeSocketFloat)
        _ = create_socket(interface, "Normal Intensity", NodeSocketFloat)
        _ = create_socket(interface, "Tint Color", NodeSocketColor)

    def create_nodes(self) -> None:
        if self.node_tree is None:
            return
        nodes = self.node_tree.nodes
        input = create_node(nodes, 0, 0, NodeGroupInput)
        output = create_node(nodes, 0, 0, NodeGroupOutput)

        bsdf = create_node(nodes, 0, 0, ShaderNodeBsdfPrincipled)

        math = create_node(nodes, 0, 0, ShaderNodeMath)
        math.operation = "MULTIPLY"

        geometry = create_node(nodes, 0, 0, ShaderNodeNewGeometry)
        mix_shader = create_node(nodes, 0, 0, ShaderNodeMixShader)
        transparent = create_node(nodes, 0, 0, ShaderNodeBsdfTransparent)

        normal_map = create_node(nodes, 0, 0, ShaderNodeNormalMap)

        normalize = create_node(nodes, 0, 0, ShaderNodeGroup)
        normalize.node_tree = cast(ShaderNodeTree, NormNormalize().node_tree)
        assign_value(normalize, 1, 1.0)

        mix = create_node(nodes, 0, 0, ShaderNodeMix)
        mix.data_type = "RGBA"
        mix.blend_type = "MULTIPLY" #TODO: Figure out blending logic
        assign_value(mix, 0, 1.0)

        links = self.node_tree.links
        create_link(links, input, mix, 0, 6)
        create_link(links, input, mix, 7, 7)
        create_link(links, input, normalize, 2, 0)
        create_link(links, normalize, normal_map, 0, 1)
        create_link(links, input, normal_map, 6, 0)
        create_link(links, normal_map, bsdf, 0, 6)
        create_link(links, input, math, 1, 0)
        create_link(links, input, math, 4, 1)
        create_link(links, mix, bsdf, 2, 0)
        create_link(links, math, bsdf, 0, 4)
        create_link(links, input, bsdf, 2, 2)
        create_link(links, input, bsdf, 3, 2)
        create_link(links, input, bsdf, 5, 1)
        create_link(links, geometry, mix_shader, 6, 0)
        create_link(links, bsdf, mix_shader, 0, 1)
        create_link(links, transparent, mix_shader, 0, 2)
        create_link(links, mix_shader, output, 0, 0)

File: crates/ekur_materials/src/basic/color_tint_decal.rs
/* SPDX-License-Identifier: GPL-3.0-or-later */
/* Copyright © 2026 The Halo Archive */
use anyhow::Result;
use ekur_definitions::material::MaterialPostProcessing;

use crate::utils::{f32_from_const, get_post_texture};

use crate::{ColorTintDecal, Material, ShaderType, TextureType};

pub(crate) fn handle_color_tint_decal(
    post_process: &MaterialPostProcessing,
    material: &mut Material,
) -> Result<()> {
    let mut color_decal = ColorTintDecal::default();
    get_post_texture(post_process, material, 60, TextureType::Color)?;
    get_post_texture(post_process, material, 16, TextureType::Normal)?;
    color_decal.opacity = f32_from_const(material, 12)?;
    color_decal.roughness = f32_from_const(material, 52)?;
    color_decal.metallic = f32_from_const(material, 56)?;
    color_decal.normal_intensity = f32_from_const(material, 48)?;
    color_decal.tint_color = (
        f32_from_const(material, 0)?,
        f32_from_const(material, 4)?,
        f32_from_const(material, 8)?,
    );
    material.color_tint_decal = Some(color_decal);
    material.shader_type = ShaderType::ColorTintDecal;
    Ok(())
}

pub(crate) fn handle_color_notint_decal(
    post_process: &MaterialPostProcessing,
    material: &mut Material,
) -> Result<()> {
    let mut color_decal = ColorTintDecal::default();
    get_post_texture(post_process, material, 44, TextureType::Color)?;
    get_post_texture(post_process, material, 4, TextureType::Normal)?;
    color_decal.opacity = f32_from_const(material, 0)?;
    color_decal.roughness = f32_from_const(material, 36)?;
    color_decal.metallic = f32_from_const(material, 40)?;
    color_decal.normal_intensity = f32_from_const(material, 32)?;
    color_decal.tint_color = (1.0, 1.0, 1.0);
    material.color_tint_decal = Some(color_decal);
    material.shader_type = ShaderType::ColorTintDecal;
    Ok(())
}
